clients/orderbook/src/worker.rs

```diff
@@ -0,0 +1,1289 @@
+		set_bit_field(&mut summary.bitflags, bit_index as u16);
+		// println!("Signature: {:?}",summary.aggregate_signature.unwrap().0);
+		// println!("Signing key: {:?}",signing_key.0);
+		assert_eq!(initial_summary, summary.sign_data());
+		assert!(bls_primitives::crypto::bls_ext::verify(
+			&signing_key,
+			&initial_summary,
+			&signature
+		));
+		// send it to runtime
+		if let Err(_) = self
+			.runtime
+			.runtime_api()
+			.submit_snapshot(&BlockId::number(self.last_finalized_block.into()), summary)
+			.expect("Something went wrong with the submit_snapshot runtime api; qed.")
+		{
+			error!(target:"orderbook","📒 Failed to submit snapshot to runtime");
+			return Err(Error::FailedToSubmitSnapshotToRuntime)
+		}
+		Ok(())
+	}
+
+	pub fn handle_action(&mut self, action: &ObMessage) -> Result<(), Error> {
+		info!(target:"orderbook","📒 Processing action: {:?}", action);
+		match action.action.clone() {
+			// Get Trie here itself and pass to required function
+			// No need to change Test cases
+			UserActions::Trade(trades) => {
+				let mut memory_db = self.memory_db.write();
+				let mut working_state_root = self.working_state_root.write();
+				let mut trie = Self::get_trie(&mut memory_db, &mut working_state_root);
+
+				for trade in trades {
+					process_trade(&mut trie, trade)?
+				}
+				// Commit the trie
+				trie.commit();
+			},
+			UserActions::Withdraw(withdraw) => self.process_withdraw(withdraw, action.stid)?,
+			UserActions::BlockImport(num) => self.handle_blk_import(num)?,
+			UserActions::Snapshot => self.snapshot(action.stid)?,
+		}
+		self.latest_stid = action.stid;
+		// Multicast the message to other peers
+		self.gossip_engine.gossip_message(topic::<B>(), action.encode(), true);
+		Ok(())
+	}
+
+	// Checks if we need to sync the orderbook state before processing the messages.
+	pub async fn check_state_sync(&mut self) -> Result<(), Error> {
+		// X->Y sync: Ask peers to send the missed stid
+		if !self.known_messages.is_empty() {
+			// Collect all known stids
+			let mut known_stids = self.known_messages.keys().collect::<Vec<&u64>>();
+			known_stids.sort_unstable(); // unstable is fine since we know stids are unique
+							 // if the next best known stid is not available then ask others
+			if *known_stids[0] != self.last_snapshot.read().state_change_id.saturating_add(1) {
+				// Ask other peers to send us the requests stids.
+				let message = GossipMessage::WantStid(
+					self.last_snapshot.read().state_change_id,
+					*known_stids[0],
+				);
+				let mut peers = self
+					.gossip_validator
+					.peers
+					.read()
+					.clone()
+					.iter()
+					.cloned()
+					.collect::<Vec<PeerId>>();
+				let mut fullnodes = self
+					.gossip_validator
+					.fullnodes
+					.read()
+					.clone()
+					.iter()
+					.cloned()
+					.collect::<Vec<PeerId>>();
+				peers.append(&mut fullnodes);
+				// TODO: Should we even send it out to everyone we know?
```

Comment:
Still pending

Reply:
It's not pending, these are something to test in testnet and then tweak.